innvestigate/analyzer/reverse_map.py

```diff
@@ -0,0 +1,335 @@
+        """
+        outs = args
+
+        if reversed_outs is None:
+            reversed_outs = outs
+
+        if len(self.layer_next) > 1:
+            #TODO is this addition correct?
+            ret = keras_layers.Add(dtype=tf.float32)([r for r in reversed_outs])
+        elif len(self.input_shape) > 1:
+            ret = [reversed_outs for i in self.input_shape]
+            ret = tf.keras.layers.concatenate(ret, axis=1)
+        else:
+            ret = reversed_outs
+        return ret
+
+
+class GradientReplacementLayer(ReplacementLayer):
+    def __init__(self, *args, **kwargs):
+        super(GradientReplacementLayer, self).__init__(*args, **kwargs)
+
+    def wrap_hook(self, ins, neuron_selection):
+        #print("WRAP ", ins.shape, neuron_selection.shape)
+        with tf.GradientTape(persistent=True) as tape:
+            tape.watch(ins)
+            outs = self.apply(ins, neuron_selection)
+
+        return outs, tape
+
+    def explain_hook(self, ins, reversed_outs, args):
+        outs, tape = args
+
+        if reversed_outs is None:
+            reversed_outs = outs
+
+        # correct number of outs
+        if len(self.layer_next) > 1:
+            outs = [outs for l in self.layer_next]
+
+        if len(self.layer_next) > 1:
+            #print(self.name, np.shape(outs), np.shape(reversed_outs))
+            #TODO: is this correct?
+            keras_layers.Add(dtype=tf.float32)([tape.gradient(o, ins, output_gradients=r) for o, r in zip(outs, reversed_outs)])
+            #raise ValueError("This basic function is not defined for layers with multiple children")
+        if len(self.input_shape) > 1:
+            ret = [tape.gradient(outs, i, output_gradients=reversed_outs) for i in ins]
+        else:
+            ret = tape.gradient(outs, ins, output_gradients=reversed_outs)
+        return ret
+
+
+def reverse_map(
+    #Alternative to kgraph.reverse_model.
+        model,
+        reverse_mappings,
+        default_reverse_mapping,
+        stop_mapping_at_tensors,
+        verbose=False):
+
+    #TODO: verbose
+    #TODO: HeadMapping
+    #TODO this is just the basic core. Add full functionality of kgraph.reverse_model
+
+    #build model that is to be analyzed
+    layers = kgraph.get_model_layers(model)
+    stop_mapping_at_tensors = [x.name.split(":")[0] for x in stop_mapping_at_tensors]
+
+    # set all replacement layers
+    replacement_layers = []
+    for layer in layers:
+        if not layer.name in stop_mapping_at_tensors:
+            layer_next = []
+            wrapper_class = reverse_mappings(layer)
+            if wrapper_class is None:
+                wrapper_class = default_reverse_mapping(layer)
+
+            if not issubclass(wrapper_class, ReplacementLayer):
+                raise ValueError("Reverse Mappings should be an instance of ReplacementLayer")
+
+            replacement_layers.append(wrapper_class(layer, layer_next))
+
+    #connect graph structure
+    for layer in replacement_layers:
```

Comment:
In Keras one layer can be applied to several inputs. Does the code in this case still work? It would be great to have comment on how that case is handled.

Reply:
Do you mean the case for e.g. concatenate layers? The code considers this case by aggregating inputs for each layer before applying the forward pass. see try_apply function in innvestigate/analyzer/reverse_map.py:146.